src/Http/Routing/src/EndpointRoutingMiddleware.cs

```diff
@@ -107,12 +112,21 @@ private Task SetRoutingAndContinue(HttpContext httpContext)
                 Write(_diagnosticListener, httpContext);
             }
 
-            Log.MatchSuccess(_logger, endpoint);
-
-            if (_logger.IsEnabled(LogLevel.Debug)
-                && endpoint.Metadata.GetMetadata<FallbackMetadata>() is not null)
+            if (_logger.IsEnabled(LogLevel.Debug) || _metrics.MatchSuccessCounterEnabled)
             {
-                Log.FallbackMatch(_logger, endpoint);
+                var isFallback = endpoint.Metadata.GetMetadata<FallbackMetadata>() is not null;
+
+                Log.MatchSuccess(_logger, endpoint);
+
+                if (isFallback)
+                {
+                    Log.FallbackMatch(_logger, endpoint);
+                }
+
+                // It shouldn't be possible for a route to be matched via the route matcher and not have a route.
+                // Just in case, add a special (unknown) value as the route tag to metrics.
+                var route = endpoint.Metadata.GetMetadata<IRouteDiagnosticsMetadata>()?.Route ?? "(unknown)";
```

Comment:
This could be the case if some other middleware sets the endpoint, which they should be able to do?

Reply:
This is the endpoint-matching middleware, not the endpoint execution middleware. Endpoints from other middleware won't reach here.

The endpoint here has to be one returned by the matcher. `IRouteDiagnosticsMetadata` metadata is automatically added if the route was created with `RouteEndpointBuilder`. If the endpoint is created another way then someone could add metadata themselves.

Reply:
Fair enough, I always confuse between the two 😄